rclcpp_action/src/server.cpp

```diff
@@ -232,9 +232,13 @@ ServerBase::execute_goal_request_received()
   GoalUUID uuid = get_goal_id_from_goal_request(message.get());
   convert(uuid, &goal_info);
 
+  lock.unlock();
+
   // Call user's callback, getting the user's response and a ros message to send back
   auto response_pair = call_handle_goal_callback(uuid, message);
 
+  lock.lock();
+
```

Comment:
These type of lock/unlock sequences are often problematic.  It often leads to race conditions where two threads come in here simultaneously, then race to the `lock.lock()`.  Only one will win (and the other will have to wait), but what if T1 comes in first, calls the callback, but then T2 is the one that reacquires the lock first?  Is that situation OK?

Can you explain more about the original deadlock situation?  I'm wondering if there are other solutions we can use here.

Reply:
@clalancette 

> if T1 comes in first, calls the callback, but then T2 is the one that reacquires the lock first? Is that situation OK?

There are 3 `call_**_callback` functions that do not call any `rcl` functions that should be protected by the lock (`pimpl_->reentrant_mutex_`)

- `call_handle_goal_callback`: this uses only local variables
- `call_goal_accepted_callback`: 
- `call_handle_cancel_callback`: 
  - These will access to `goal_handles_`, but also protected by `goal_handles_mutex_`

So, I believe these three functions are threadsafe as long as user callbacks are threadsafe which we don't need to worry about. 

> Can you explain more about the original deadlock situation? I'm wondering if there are other solutions we can use here.

I think explaining with my test code is the best way. The user code has its own mutex which is used in the callback functions and also is used in their working thread which calls the succeed function. 

https://github.com/ros2/rclcpp/blob/938f973c99542a52d413e3e3b8d083acec78d6b7/rclcpp_action/test/test_server.cpp#L994-L1003

- user code: lock user mutex and call succeed (eventually lock rclcpp_action)  L997
- rclcpp_action: lock rclcpp_action mutex and call callbacks (lock user mutex) L1002

Reply:
@clalancette 

I think the most complicated case is the following,

in condition, the goals will be accepted always and simply only two threads.

1. (T-1) `goal_request_ready_` ready, then call `execute_goal_request_received` with `reentrant_mutex_` locked
2. (T-1) unlock `reentrant_mutex_`, then it will call `call_handle_goal_callback` (accepted)
3. (T-2) `goal_request_ready_` ready, then call `execute_goal_request_received` with `reentrant_mutex_` locked
4. (T-2) unlock `reentrant_mutex_`, then it will call `call_handle_goal_callback` (accepted)
5. (T-2) lock `reentrant_mutex_`, call `rcl_action_send_goal_response` and the rest.
6. (T-2) unlock `reentrant_mutex_`.
7. (T-1) lock `reentrant_mutex_`, call `rcl_action_send_goal_response` and the rest.
8. (T-1) unlock `reentrant_mutex_`.

I believe this is the new prblem introduced by this fix, because goal order between acceptance and execution is not consistent. (it is really rare, but possible.)

Reply:
If that order should be guranteed, we can put another mutex which is only used at the beginning of `execute_goal_request_received`.

Reply:
@fujitatomoya
I don't think the potential issue is a critical issue because this only happens when the action server is executed by a multithread executor.

Reply:
yes, obviously that is one of the case. I think the `reentrant_mutex_` is to support exclusive access for `execute_goal_request_received` since it was considered as issue. IMO, it would be better not to change the behavior. for example, if user application expects `first comes, first accepted & executed`, which is really common I guess. this cannot be guaranteed by the fix. I'd like to hear from others.

Reply:
I added another lock to keep the behavior.